### PR TITLE
[PB-1850]: Feat/Workspaces team management

### PR DESCRIPTION
**Pull request by apsantiso**
This PR implements workspaces team management.

**There are some questions that might surge:**

- Usecases not checking permissions: most usecases requires manager or workspace owner role, the roles are being checked properly by the guard set in the controller. To maintain clean usecases, I decided not to re-check again those things. 
- Usecases checking if team is a default team: default teams of workspaces are not editable, we do not want to be able to edit them.

### CODE REVIEW

File: test/fixtures.ts [hunk 1]
```diff
@@ -311,6 +312,33 @@ export const newWorkspaceTeam = (params?: {
...
+    memberId: params?.memberId || v4(),
+    createdAt: randomCreatedAt,
+    updatedAt: new Date(
+      randomDataGenerator.date({
+        min: randomCreatedAt,
+      }),
+    ),
+  });
+
+  params?.attributes &&
```
Comment (sg-gs) [inline test/fixtures.ts:334]: Add a test for this fixture, especially for this dynamic keys assignation

File: src/modules/workspaces/repositories/team.repository.ts [hunk 2]
```diff
@@ -20,6 +20,10 @@ export class SequelizeWorkspaceTeamRepository {
     private teamUserModel: typeof WorkspaceTeamUserModel,
   ) {}
 
+  createTransaction(): Promise<Transaction> {
```
Comment (sg-gs) [inline src/modules/workspaces/repositories/team.repository.ts:23]: We do not want to use transactions in order to scale, therefore, this function should be removed

File: src/modules/workspaces/repositories/team.repository.ts [hunk 3]
```diff
@@ -92,6 +104,17 @@ export class SequelizeWorkspaceTeamRepository {
...
   }
 
+  async removeMemberFromTeam(
+    teamId: WorkspaceTeamAttributes['id'],
+    memberId: User['uuid'],
+    transaction?: Transaction,
+  ): Promise<void> {
+    await this.teamUserModel.destroy({
+      where: { teamId, memberId },
+      transaction,
```
Comment (sg-gs) [inline src/modules/workspaces/repositories/team.repository.ts:114]: Remove the transaction usage

File: src/modules/workspaces/workspaces.controller.ts [hunk 4]
```diff
@@ -152,11 +155,136 @@ export class WorkspacesController {
...
+  @ApiOkResponse({
+    description: 'Team has been edited',
+  })
+  @UseGuards(WorkspaceGuard)
+  @WorkspaceRequiredAccess(AccessContext.TEAM, WorkspaceRole.MANAGER)
+  async editTeam(
+    @Param('teamId') teamId: WorkspaceTeamAttributes['id'],
+    @Body() editTeamBody: EditTeamDto,
+  ) {
+    if (!teamId || !isUUID(teamId)) {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.controller.ts:188]: Do we have some validator for this? I remember that we had something to validate UUIDs and is used on the DTOs so we may be able to use it here.

File: src/modules/workspaces/workspaces.controller.ts [hunk 5]
```diff
@@ -152,11 +155,136 @@ export class WorkspacesController {
...
+    summary: 'Deletes a team in a workspace',
+  })
+  @ApiBearerAuth()
+  @ApiOkResponse({
+    description: 'Deleted team',
+  })
+  @UseGuards(WorkspaceGuard)
+  @WorkspaceRequiredAccess(AccessContext.WORKSPACE, WorkspaceRole.OWNER)
+  async deleteTeam(@Param('teamId') teamId: WorkspaceTeamAttributes['id']) {
+    if (!teamId || !isUUID(teamId)) {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.controller.ts:206]: Same here

File: src/modules/workspaces/workspaces.controller.ts [hunk 6]
```diff
@@ -152,11 +155,136 @@ export class WorkspacesController {
...
   @UseGuards(WorkspaceGuard)
-  @WorkspaceRequiredAccess(AccessContext.TEAM, WorkspaceRole.MEMBER)
-  async getTeamMembers() {
-    throw new NotImplementedException();
+  @WorkspaceRequiredAccess(AccessContext.TEAM, WorkspaceRole.MANAGER)
+  async addUserToTeam(
+    @Param('teamId') teamId: WorkspaceTeamAttributes['id'],
+    @Param('userUuid') memberId: UserAttributes['uuid'],
+  ) {
+    if (!teamId || !isUUID(teamId)) {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.controller.ts:228]: And here

File: src/modules/workspaces/workspaces.controller.ts [hunk 7]
```diff
@@ -152,11 +155,136 @@ export class WorkspacesController {
...
+      throw new BadRequestException('Invalid team id');
+    }
+
+    if (!memberId || !isUUID(memberId)) {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.controller.ts:232]: And here

File: src/modules/workspaces/workspaces.controller.ts [hunk 8]
```diff
@@ -152,11 +155,136 @@ export class WorkspacesController {
...
+  @ApiOkResponse({
+    description: 'User removed from team',
+  })
+  @UseGuards(WorkspaceGuard)
+  @WorkspaceRequiredAccess(AccessContext.TEAM, WorkspaceRole.MANAGER)
+  async removeUserFromTeam(
+    @Param('teamId') teamId: WorkspaceTeamAttributes['id'],
+    @Param('userUuid') memberId: UserAttributes['uuid'],
+  ) {
+    if (!teamId || !isUUID(teamId)) {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.controller.ts:259]: And here

File: src/modules/workspaces/workspaces.controller.ts [hunk 9]
```diff
@@ -152,11 +155,136 @@ export class WorkspacesController {
...
+      throw new BadRequestException('Invalid Team id');
+    }
+
+    if (!memberId || !isUUID(memberId)) {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.controller.ts:263]: And here

File: src/modules/workspaces/workspaces.controller.ts [hunk 10]
```diff
@@ -152,11 +155,136 @@ export class WorkspacesController {
...
+
+    return newTeamMember;
+  }
+
+  @Delete('/teams/:teamId/user/:userUuid')
+  @ApiOperation({
+    summary: 'Remove user from team',
+  })
+  @ApiBearerAuth()
+  @ApiParam({ name: 'teamId', type: String, required: true })
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.controller.ts:248]: The required: true does remove the necessity of the `if (!teamId)`, isn't it?

File: src/modules/workspaces/workspaces.controller.ts [hunk 11]
```diff
@@ -152,11 +155,136 @@ export class WorkspacesController {
...
+  @ApiOkResponse({
+    description: 'Team manager changed',
+  })
+  @UseGuards(WorkspaceGuard)
+  @WorkspaceRequiredAccess(AccessContext.WORKSPACE, WorkspaceRole.OWNER)
+  async changeTeamManager(
+    @Param('teamId') teamId: WorkspaceTeamAttributes['id'],
+    @Body('managerId') managerId: UserAttributes['uuid'],
+  ) {
+    if (!managerId || !isUUID(managerId)) {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.controller.ts:284]: And here

File: src/modules/workspaces/workspaces.usecase.spec.ts [hunk 12]
```diff
@@ -949,4 +950,441 @@ describe('WorkspacesUsecases', () => {
...
+
+        await expect(
+          service.createTeam(user, workspace.id, {
+            name: 'test',
+            managerId: '',
+          }),
+        ).rejects.toThrow(BadRequestException);
+      });
+
+      it('When maximum teams in a workspace is still not reached, then create workspace', async () => {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.spec.ts:986]: In the workspace lifecycle, a workspace comes first. How is it possible to check the maximum number of teams in a workspace of a non-existent workspace?

I assume that the legend is wrong by reading the code and it should finish with _then the team is created succesfully_

File: src/modules/workspaces/workspaces.usecase.spec.ts [hunk 13]
```diff
@@ -949,4 +950,441 @@ describe('WorkspacesUsecases', () => {
...
+        const editTeamDto = { name: 'Updated Team' };
+
+        jest.spyOn(teamRepository, 'getTeamById').mockResolvedValue(null);
+
+        await expect(service.editTeamData(teamId, editTeamDto)).rejects.toThrow(
+          BadRequestException,
+        );
+      });
+
+      it('When team is default company of a workspace, then it should throw', async () => {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.spec.ts:1029]: _When the team to update is the default one of a workspace_ could be a better legend

File: src/modules/workspaces/workspaces.usecase.spec.ts [hunk 14]
```diff
@@ -949,4 +950,441 @@ describe('WorkspacesUsecases', () => {
...
+
+        jest.spyOn(teamRepository, 'getTeamById').mockResolvedValue(team);
+        jest.spyOn(workspaceRepository, 'findOne').mockResolvedValue(workspace);
+
+        await expect(
+          service.editTeamData(team.id, editTeamDto),
+        ).rejects.toThrow(BadRequestException);
+      });
+
+      it('When team is being updated, then it update the respective fields', async () => {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.spec.ts:1047]: _When the team is part of the workspace and is not the default one, then the update is done_ could be a better legend

File: src/modules/workspaces/workspaces.usecase.ts [hunk 15]
```diff
@@ -513,6 +521,145 @@ export class WorkspacesUsecases {
...
+    return members;
+  }
+
+  async editTeamData(teamId: WorkspaceTeam['id'], editTeamDto: EditTeamDto) {
+    const [team, workspace] = await Promise.all([
+      this.teamRepository.getTeamById(teamId),
+      this.workspaceRepository.findOne({ defaultTeamId: teamId }),
+    ]);
+
+    if (!team || workspace?.isDefaultTeam(team)) {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.ts:535]: The `workspace?.isDefaultTeam(team)` check seems to be irrelevant. If the query finds it, it's the default team, if not, could be due to 2 reasons: 
- The workspace is not the team's, thus, we need to prevent the update
- The workspace is the team's one

If instead of looking at the workspace by the default team you search it by the `team.workspaceId`, things are easier to check:

```
const team = await this.teamRepository.getTeamById(teamId);
const workspace = await this.workspaceRepository.findById(team.workspaceId);

if (!workspace) {
  // someone is doing something nasty -> Forbidden
}

if (!team || workspace.isDefaultTeam(team)) {
  // bad request
}

/*
If you are here, the team is part of the workspace but it is not the default one, safe to continue. Non-existent workspaces or teams and non-related workspaces cases are handled properly
*/
```

File: src/modules/workspaces/workspaces.usecase.spec.ts [hunk 16]
```diff
@@ -949,4 +950,441 @@ describe('WorkspacesUsecases', () => {
...
+        const memberId = 'member-uuid';
+
+        jest.spyOn(teamRepository, 'getTeamById').mockResolvedValue(null);
+
+        await expect(service.addMemberToTeam(teamId, memberId)).rejects.toThrow(
+          BadRequestException,
+        );
+      });
+
+      it('When team is default company of a workspace, then it should throw', async () => {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.spec.ts:1080]: _When the team to update is the default one of a workspace_ could be a better legend

File: src/modules/workspaces/workspaces.usecase.spec.ts [hunk 17]
```diff
@@ -949,4 +950,441 @@ describe('WorkspacesUsecases', () => {
...
+        await service.removeMemberFromTeam(team.id, member.uuid);
+
+        expect(teamRepository.removeMemberFromTeam).toHaveBeenCalledWith(
+          team.id,
+          member.uuid,
+        );
+      });
+    });
+
+    describe('changeTeamManager', () => {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.spec.ts:1273]: What happens if the team manager that is going to be changed is the manager of the default team? (In this case, the owner?

File: src/modules/workspaces/workspaces.usecase.spec.ts [hunk 18]
```diff
@@ -949,4 +950,441 @@ describe('WorkspacesUsecases', () => {
...
+        const teamId = 'team-id';
+
+        jest.spyOn(teamRepository, 'getTeamById').mockResolvedValue(null);
+
+        await expect(service.deleteTeam(teamId)).rejects.toThrow(
+          BadRequestException,
+        );
+      });
+
+      it('When team is default team, then it should throw', async () => {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.spec.ts:1361]: _When the team to update is the default one of a workspace_ could be a better legend

File: src/modules/workspaces/workspaces.usecase.ts [hunk 19]
```diff
@@ -513,6 +521,145 @@ export class WorkspacesUsecases {
...
+  async removeMemberFromTeam(
+    teamId: WorkspaceTeam['id'],
+    memberId: User['uuid'],
+  ) {
+    const [team, workspace] = await Promise.all([
+      this.teamRepository.getTeamById(teamId),
+      this.workspaceRepository.findOne({ defaultTeamId: teamId }),
+    ]);
+
+    if (!team || workspace?.isDefaultTeam(team)) {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.ts:551]: Same issue as with `editTeamData`

File: src/modules/workspaces/workspaces.usecase.ts [hunk 20]
```diff
@@ -513,6 +521,145 @@ export class WorkspacesUsecases {
...
+    await this.teamRepository.removeMemberFromTeam(teamId, memberId);
+  }
+
+  async addMemberToTeam(teamId: WorkspaceTeam['id'], memberId: User['uuid']) {
+    const [team, workspace] = await Promise.all([
+      this.teamRepository.getTeamById(teamId),
+      this.workspaceRepository.findOne({ defaultTeamId: teamId }),
+    ]);
+
+    if (!team || workspace?.isDefaultTeam(team)) {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.ts:582]: Same issue as with `editTeamData`. As this check seems to be repetitive, I suggest to create a function just for this

File: src/modules/workspaces/workspaces.usecase.ts [hunk 21]
```diff
@@ -513,6 +521,145 @@ export class WorkspacesUsecases {
...
+  async changeTeamManager(
+    teamId: WorkspaceTeam['id'],
+    managerId: User['uuid'],
+  ) {
+    const [team, workspace] = await Promise.all([
+      this.teamRepository.getTeamById(teamId),
+      this.workspaceRepository.findOne({ defaultTeamId: teamId }),
+    ]);
+
+    if (!team || workspace?.isDefaultTeam(team)) {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.ts:626]: Same here

File: src/modules/workspaces/workspaces.usecase.ts [hunk 22]
```diff
@@ -513,6 +521,145 @@ export class WorkspacesUsecases {
...
+    await this.teamRepository.updateById(team.id, { managerId });
+  }
+
+  async deleteTeam(teamId: WorkspaceTeam['id']) {
+    const [team, workspace] = await Promise.all([
+      this.teamRepository.getTeamById(teamId),
+      this.workspaceRepository.findOne({ defaultTeamId: teamId }),
+    ]);
+
+    if (!team || workspace?.isDefaultTeam(team)) {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.ts:656]: Same here

**Review by sg-gs — CHANGES_REQUESTED** [review]

File: src/modules/workspaces/workspaces.controller.ts (hunk 10 above)
Comment (apsantiso) [reply src/modules/workspaces/workspaces.controller.ts:234]: It is for openAPI specifications, it does not affect the logic in the controller

File: src/modules/workspaces/workspaces.usecase.spec.ts (hunk 17 above)
Comment (apsantiso) [reply src/modules/workspaces/workspaces.usecase.spec.ts:1291]: It won't proceed if deault team = true

File: src/modules/workspaces/workspaces.usecase.spec.ts [hunk 23]
```diff
@@ -949,4 +951,447 @@ describe('WorkspacesUsecases', () => {
...
+        const teamId = 'team-id';
+
+        jest.spyOn(teamRepository, 'getTeamById').mockResolvedValue(null);
+
+        await expect(
+          service.getAndValidateNonDefaultTeamWorkspace(teamId),
+        ).rejects.toThrow(BadRequestException);
+      });
+
+      it('When team is default company of the workspace, then it should throw', async () => {
```
Comment (sg-gs) [inline src/modules/workspaces/workspaces.usecase.spec.ts:1029]: _When team is the default one of the workspace_ could be clearer, the usage of the word _company_ here seems misleading in this context

**Review by sg-gs — APPROVED** [review]

File: src/modules/workspaces/repositories/team.repository.ts [hunk 24]
```diff
@@ -3,7 +3,7 @@ import { InjectModel } from '@nestjs/sequelize';
 import { UserModel } from '../../user/user.model';
 import { User } from '../../user/user.domain';
 import { WorkspaceAttributes } from '../attributes/workspace.attributes';
-import { Sequelize, Transaction } from 'sequelize';
+import { Sequelize } from 'sequelize';
```
[the remark this answers is not in the document]
Comment (apsantiso) [reply src/modules/workspaces/repositories/team.repository.ts:6]: Hi. It is used to make this query   

```  async getTeamsAndMembersCountByWorkspace(
    workspaceId: WorkspaceAttributes['id'],
  ): Promise<{ team: WorkspaceTeam; membersCount: number }[]> {
    const teams = await this.teamModel.findAll({
      where: { workspaceId },
      include: [
        {
          model: WorkspaceTeamUserModel,
          attributes: [],
        },
      ],
      attributes: {
        include: [
          [
            Sequelize.fn('COUNT', Sequelize.col('teamUsers.id')),
            'membersCount',
          ],
        ],
      },
      group: ['WorkspaceTeamModel.id'],
    });

    return teams.map((team) => ({
      team: this.toDomain(team),
      membersCount: parseInt(team.dataValues?.membersCount),
    }));
  }```
